services/src/main/java/org/exoplatform/officeonline/WOPIService.java

```diff
@@ -163,6 +196,98 @@ public WOPIService(SessionProviderService sessionProviders,
+    fileExtensions.put("application/vnd.openxmlformats-officedocument.presentationml.slideshow", "ppsx");
+    fileExtensions.put("application/vnd.ms-powerpoint.addin.macroEnabled.12", "ppam");
+    fileExtensions.put("application/vnd.ms-powerpoint.presentation.macroEnabled.12", "pptm");
+    fileExtensions.put("application/vnd.ms-powerpoint.template.macroEnabled.12", "potm");
+    fileExtensions.put("application/vnd.ms-powerpoint.slideshow.macroEnabled.12", "ppsm");
+  }
+
+  /**
+   * Put file.
+   *
+   * @param config the config
+   * @param data the data
+   * @throws OfficeOnlineException the office online exception
+   * @throws SizeMismatchException 
+   */
+  public void putFile(EditorConfig config, String lockId, InputStream data) throws OfficeOnlineException,
+                                                                            LockMismatchException,
+                                                                            SizeMismatchException,
+                                                                            RepositoryException {
+    Node node = nodeByUUID(config.getFileId(), config.getWorkspace());
+    try {
+      if (canEditDocument(node)) {
+        Node content = node.getNode(JCR_CONTENT);
+        if (!node.isLocked()) {
+          long size = content.getProperty(JCR_DATA).getLength();
+          if (size != 0) {
+            throw new SizeMismatchException("File is unlocked and size isn't equal to 0.", "");
+          }
+        } else {
+          String lockToken = node.getLock().getLockToken();
+          if (!lockToken.equals(lockId)) {
+            throw new LockMismatchException("Given lock is different from the file lock", lockToken);
+          }
+        }
+
+        content.setProperty(JCR_DATA, data);
+        Calendar editedTime = Calendar.getInstance();
+        content.setProperty(JCR_LAST_MODIFIED, editedTime);
+        if (content.hasProperty(EXO_DATE_MODIFIED)) {
+          content.setProperty(EXO_DATE_MODIFIED, editedTime);
+        }
+        if (content.hasProperty(EXO_LAST_MODIFIED_DATE)) {
+          content.setProperty(EXO_LAST_MODIFIED_DATE, editedTime);
+        }
+        if (node.hasProperty(EXO_LAST_MODIFIED_DATE)) {
+          node.setProperty(EXO_LAST_MODIFIED_DATE, editedTime);
+        }
+        if (node.hasProperty(EXO_DATE_MODIFIED)) {
+          node.setProperty(EXO_DATE_MODIFIED, editedTime);
+        }
+        if (node.hasProperty(EXO_LAST_MODIFIER)) {
+          node.setProperty(EXO_LAST_MODIFIER, config.getUserId());
+        }
+        if (data != null) {
+          try {
+            data.close();
+          } catch (IOException e) {
+            LOG.error("Error closing data stream. FileID:" + config.getFileId());
+          }
+        }
+      } else {
+        throw new PermissionDeniedException("Cannnot update file. Permission denied");
+      }
+    } catch (RepositoryException e) {
+      LOG.error("Error while putFile. {}", e.getMessage());
```

Comment:
Talk to a human (admin) in warn/error messages. In debug/trace it's still can be a developer :)